src/transform/src/attribute/cardinality.rs

```diff
@@ -0,0 +1,646 @@
+
+//! Definition and helper structs for the [`Cardinality`] attribute.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use mz_expr::{
+    BinaryFunc, Id, JoinImplementation, MirRelationExpr, MirScalarExpr, TableFunc, UnaryFunc,
+    VariadicFunc,
+};
+use mz_ore::cast::CastLossy;
+use mz_repr::GlobalId;
+
+use mz_repr::explain::ExprHumanizer;
+use ordered_float::OrderedFloat;
+
+use crate::attribute::subtree_size::SubtreeSize;
+use crate::attribute::unique_keys::UniqueKeys;
+use crate::attribute::{Attribute, DerivedAttributes, Env, RequiredAttributes};
+use crate::symbolic::SymbolicExpression;
+
+use super::Arity;
+
+/// Compute the estimated cardinality of each subtree of a [MirRelationExpr] from the bottom up.
+#[allow(missing_debug_implementations)]
+pub struct Cardinality {
+    /// Environment of computed values for this attribute
+    env: Env<Self>,
+    /// A vector of results for all nodes in the visited tree in
+    /// post-visit order
+    pub results: Vec<SymExp>,
+    /// A factorizer for generating appropriating scaling factors
+    pub factorize: Box<dyn Factorizer + Send + Sync>,
+}
+
+impl Default for Cardinality {
+    fn default() -> Self {
+        Cardinality {
+            env: Env::default(),
+            results: Vec::new(),
+            factorize: Box::new(WorstCaseFactorizer {
+                cardinalities: BTreeMap::new(),
+            }),
+        }
+    }
+}
+
+/// The variables used in symbolic expressions representing cardinality
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum FactorizerVariable {
+    /// The total cardinality of a given global id
+    Id(GlobalId),
+    /// The inverse of the number of distinct keys in the index held in the given column, i.e., 1/|# of distinct keys|
+    ///
+    /// TODO(mgree): need to correlate this back to a given global table, to feed in statistics (or have a sub-attribute for collecting this)
+    Index(usize),
+    /// An unbound local or other unknown quantity
+    Unknown,
+}
+
+/// SymbolicExpressions specialized to factorizer variables
+pub type SymExp = SymbolicExpression<FactorizerVariable>;
+
+/// A `Factorizer` computes selectivity factors
+pub trait Factorizer {
+    /// Compute selectivity for the flat map of `tf`
+    fn flat_map(&self, tf: &TableFunc, input: &SymExp) -> SymExp;
+    /// Computes selectivity of the predicate `expr`, given that `indexed_columns` are indexed
+    ///
+    /// The result should be in the range (0, 1.0]
+    fn predicate(&self, expr: &MirScalarExpr, indexed_columns: &BTreeSet<usize>) -> SymExp;
+    /// Computes selectivity for a filter
+    fn filter(
+        &self,
+        predicates: &Vec<MirScalarExpr>,
+        keys: &Vec<Vec<usize>>,
+        input: &SymExp,
+    ) -> SymExp;
+    /// Computes selectivity for a join; the cardinality estimate for each input is paired with the keys on that input
+    ///
+    /// `indexed_columns` maps column references (with indices) to their index in `inputs`
```

Comment:
This also involves some info about unique keys right? If yes, then could you please incorporate that in the comment?